core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java

```diff
@@ -0,0 +1,157 @@
+package io.quarkus.runtime.configuration;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.function.Consumer;
+
+import javax.annotation.Priority;
+
+import org.eclipse.microprofile.config.spi.ConfigBuilder;
+import org.eclipse.microprofile.config.spi.Converter;
+import org.jboss.logging.Logger;
+
+import io.smallrye.config.SmallRyeConfigBuilder;
+
+/**
+ * This small utility class is a tool which helps populating {@link SmallRyeConfigBuilder} with
+ * {@link Converter} implementations loaded from {@link ServiceLoader}.
+ *
+ * @author Bartosz Firyn (sarxos)
+ */
+public class ConverterSupport {
+
+    private static final Logger LOG = Logger.getLogger(ConverterSupport.class);
+
+    /**
+     * A list of {@link ConverterItem} which will be loaded in static initialization. This needs
+     * to be static so we can load {@link Converter} implementations without producing reflective
+     * class build items in the deployment time. The {@link ConverterItem} instances uses generic
+     * {@link Object} type to avoid typecast errors from complier.
+     */
+    private static final List<ConverterItem<Object>> CONVERTERS = getConverters();
```

Comment:
I'm not a big fan of this pattern.

AFAICS, you don't need to do that as the `populateConverters` method is called in a `@BuildStep` so it's called when building the jar.

You could simply initialize them when in `populateConverters`. Or do I miss something?

Reply:
Hi @gsmet. In regards to your comment:

1. Can you please precise which patter did you refer to? Generic `Object` or static initialization of `ConverterItem` `List`?
2. These converter's constructors must be available for the reflection because `ServiceLoader` won't be able to initialize them and since we cannot produce reflective class build items, we need to make sure to have these converters available during static initialization because when reflected object is available in static initialization there is no need to configure reflection (this is what I learned from @dmlloyd). Loading them in `cinit` guarantees that converters are initialized because `native-image` process and initializes every class. If initialize this `ConverterItem` `List` in method body then it can only be initialized indirectly, when invocation to `populateConverters(builder)` is done from a static context of other class, which is not guaranteed at all. It's will be only assumed.

Reply:
The converter list has to be in a static field in order to make it available to the native image run time.  In general I agree with this approach.

Reply:
Ah yes, sorry, I missed we did call it again at a later stage.